File: java/year2021/month11/no145/BinaryTreePostorderTraversal.java
package year2021.month11.no145;

import java.util.ArrayList;
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

public class BinaryTreePostorderTraversal {

    public static void main(String[] args) {
        TreeNode root1 = new TreeNode(1, null, new TreeNode(2, new TreeNode(3), null));
        TreeNode root2 = null;
        TreeNode root3 = new TreeNode(1);
        TreeNode root4 = new TreeNode(1, new TreeNode(2), null);
        TreeNode root5 = new TreeNode(1, null, new TreeNode(2));
        System.out.println(postorderTraversal1(root1));
        System.out.println(postorderTraversal1(root2));
        System.out.println(postorderTraversal1(root3));
        System.out.println(postorderTraversal1(root4));
        System.out.println(postorderTraversal1(root5));
    }

    private static List<Integer> postorderTraversal1(TreeNode root) {
        TreeNode p = root;
        Deque<TreeNode> stack = new LinkedList<>();
        TreeNode pre = null;
        List<Integer> ans = new ArrayList<>();
        while (p != null || !stack.isEmpty()) {
            while (p != null) {
                stack.push(p);
                p = p.left;
            }
            p = stack.peek();
            assert p != null;
            if (p.right != null && p.right != pre) {
                p = p.right;
            } else {
                stack.pop();
                ans.add(p.val);
                pre = p;
                p = null;
            }
        }
        return ans;
    }

    public static List<Integer> postorderTraversal(TreeNode root) {
        List<Integer> ans = new ArrayList<>();
        dfs(root, ans);
        return ans;
    }

    private static void dfs(TreeNode root, List<Integer> ans) {
        if (root != null) {
            dfs(root.left, ans);
            dfs(root.right, ans);
            ans.add(root.val);
        }
    }

}

class TreeNode {

    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

/*
* 给定一个二叉树，返回它的 后序 遍历。

示例:

输入: [1,null,2,3]  
   1
    \
     2
    /
   3 

输出: [3,2,1]
进阶: 递归算法很简单，你可以通过迭代算法完成吗？

来源：力扣（LeetCode）
链接：https://leetcode-cn.com/problems/binary-tree-postorder-traversal
著作权归领扣网络所有。商业转载请联系官方授权，非商业转载请注明出处。
* */

File: go/year2025/month3/no145/BinaryTreePostorderTraversal.go
package main

import "slices"

func main() {
	root1 := &TreeNode{
		Val: 1,
		Right: &TreeNode{
			Val: 2,
			Left: &TreeNode{
				Val: 3,
			},
		},
	}
	root2 := &TreeNode{
		Val: 1,
		Left: &TreeNode{
			Val:  2,
			Left: &TreeNode{Val: 4},
			Right: &TreeNode{Val: 5,
				Left:  &TreeNode{Val: 6},
				Right: &TreeNode{Val: 7},
			},
		},
		Right: &TreeNode{
			Val: 3,
			Right: &TreeNode{Val: 8,
				Left: &TreeNode{Val: 9},
			},
		},
	}
	root3 := (*TreeNode)(nil)
	root4 := &TreeNode{Val: 1}
	println(slices.Equal(postorderTraversal(root1), recursive(root1)))
	println(slices.Equal(postorderTraversal(root2), recursive(root2)))
	println(slices.Equal(postorderTraversal(root3), recursive(root3)))
	println(slices.Equal(postorderTraversal(root4), recursive(root4)))
}

// PostOrder Traversal - Iteration, times: O(n), space: O(n)
func postorderTraversal(root *TreeNode) []int {
	ans := make([]int, 0)
	stack := make([]*TreeNode, 0)
	p := root
	var prev *TreeNode
	for len(stack) > 0 || p != nil {
		for p != nil {
			stack = append(stack, p)
			p = p.Left // 左
		}
		top := stack[len(stack)-1]
		// 需要记录前驱，因为后序遍历是左右根，所以需要判断右子树是否访问过
		if top.Right != nil && top.Right != prev { // 有右子树，并且右子树未访问
			p = top.Right // 右
			continue
		}
		stack = stack[:len(stack)-1] // 左右子树都访问过，弹出当前栈顶元素
		ans = append(ans, top.Val)   // 根
		prev = top
	}
	return ans
}

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// Recursion, times: O(n), space: O(n)
func recursive(root *TreeNode) []int {
	ans := make([]int, 0)
	dfs(root, &ans)
	return ans
}

func dfs(root *TreeNode, ans *[]int) {
	if root == nil {
		return
	}
	dfs(root.Left, ans)
	dfs(root.Right, ans)
	*ans = append(*ans, root.Val)
}

/*
给你一棵二叉树的根节点 root ，返回其节点值的 后序遍历 。

示例 1：
输入：root = [1,null,2,3]
输出：[3,2,1]

示例 2：
输入：root = [1,2,3,4,5,null,8,null,null,6,7,9]
输出：[4,6,7,5,2,9,8,3,1]

示例 3：
输入：root = []
输出：[]

示例 4：
输入：root = [1]
输出：[1]

提示：
树中节点的数目在范围 [0, 100] 内
-100 <= Node.val <= 100

进阶：递归算法很简单，你可以通过迭代算法完成吗？
*/
